cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -80,6 +80,134 @@ const FunctionDoc list_value_length_doc{
+    }
+    const auto& list_scalar = batch[0].scalar_as<BaseListScalar>();
+    if (ARROW_PREDICT_FALSE(!list_scalar.is_valid)) {
+      out->value = MakeNullScalar(
+          checked_cast<const BaseListType&>(*batch[0].type()).value_type());
+      return Status::OK();
+    }
+    auto list = list_scalar.value;
+    auto index = index_scalar.value;
+    auto len = list->length();
+    if (ARROW_PREDICT_FALSE(index < 0 ||
+                            index >= static_cast<typename IndexType::c_type>(len))) {
+      return Status::Invalid("Index ", index, " is out of bounds: should be in [0, ", len,
+                             ")");
+    }
+    ARROW_ASSIGN_OR_RAISE(out->value, list->GetScalar(index));
+    return Status::OK();
+  }
+};
+
+template <typename InListType, typename IndexType>
+void AddListElementArrayKernel(ScalarFunction* func) {
+  auto inputs = {InputType::Array(InListType::type_id),
+                 InputType::Scalar(IndexType::type_id)};
+  auto output = OutputType{ListValuesType};
+  auto sig =
+      KernelSignature::Make(std::move(inputs), std::move(output), /*is_varargs=*/false);
+  auto array_exec = ListElementArray<InListType, IndexType>;
+  ScalarKernel kernel{std::move(sig), std::move(array_exec)};
+  kernel.null_handling = NullHandling::COMPUTED_NO_PREALLOCATE;
+  kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
+  DCHECK_OK(func->AddKernel(std::move(kernel)));
+}
+
+void AddListElementArrayKernels(ScalarFunction* func) {
+  AddListElementArrayKernel<ListType, UInt8Type>(func);
+  AddListElementArrayKernel<ListType, Int8Type>(func);
+  AddListElementArrayKernel<ListType, UInt16Type>(func);
+  AddListElementArrayKernel<ListType, Int16Type>(func);
+  AddListElementArrayKernel<ListType, UInt32Type>(func);
+  AddListElementArrayKernel<ListType, Int32Type>(func);
+  AddListElementArrayKernel<ListType, UInt64Type>(func);
+  AddListElementArrayKernel<ListType, Int64Type>(func);
+
+  AddListElementArrayKernel<LargeListType, UInt8Type>(func);
+  AddListElementArrayKernel<LargeListType, Int8Type>(func);
+  AddListElementArrayKernel<LargeListType, UInt16Type>(func);
+  AddListElementArrayKernel<LargeListType, Int16Type>(func);
+  AddListElementArrayKernel<LargeListType, UInt32Type>(func);
+  AddListElementArrayKernel<LargeListType, Int32Type>(func);
+  AddListElementArrayKernel<LargeListType, UInt64Type>(func);
+  AddListElementArrayKernel<LargeListType, Int64Type>(func);
+
+  AddListElementArrayKernel<FixedSizeListType, UInt8Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int8Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, UInt16Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int16Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, UInt32Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int32Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, UInt64Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int64Type>(func);
+}
```

Comment:
The repeated code here can be simplified with a custom generator dispatcher.

Reply:
We don't need a custom generator, GenerateInteger is enough.

Reply:
Oh I see, let me try to apply the same approach @lidavidm suggested for the scalar array kernel.

Reply:
I think I have it! Thanks @edponce !